### DISCUSSION TITLE
Adding a list of wikilinks

### DISCUSSION THREAD

**Discussion by aes-gloot** [SilentVoid13/Templater#1165] [Templates Showcase]
Hi (thank-you-this-is-awesome-first-time-caller...)

A thing that I want to do, and can _almost_ do is to insert a list of wikilinks. The result that I'm trying to template looks roughly like this:

```
meeting:: Discuss world domination
- (contact:: [[Piotr H Boss|phb]])
- (contact:: [[M I Nion|mini]])
```

This is what I got, so far:
```js
<%
(async function() {
	var s = "meeting::\n";
	var p = await tp.system.prompt("Who?");
	while (p) {
		s += `- (contact:: [[${p}]])\n`
		p = await tp.system.prompt("who?")
	}
	return s;
})()
%>
```
I'd have no problem with awaiting something like an `editor:insert-wikilink` command in the middle somewhere, because I think that does exactly what I want, especially with aliases, but it would have to be in the loop somehow. 

Bonus points for restricting the link to path, but that's a stretch goal. (I already know Zeus, Taxes and Naming aren't in the meeting)

**Comment by Zachatoo** [comment]
I'll give you some resources to look into, you should be able to combine these resources in order to do what you want. Happy to provide further guidance if what I'm giving you isn't enough.

- [Suggester for getting a list of files in a folder](https://zachyoung.dev/posts/templater-snippets#suggester-for-files-in-a-specific-folder).
- [Multi-select suggester](https://zachyoung.dev/posts/templater-multi-select-suggester) user script, for being able to select multiple options from a suggester with pre-defined options (in your case, from a list of files in a folder).
- Bonus: [retrieving frontmatter from a note](https://zachyoung.dev/posts/templater-snippets#retrieve-frontmatter-from-another-note) to get the aliases from a note and make it part of the wikilink.

I like the idea of using the `editor:insert-wikilink` command, however Templater won't have a way to wait for you to finish inserting the wikilink, therefore the execution of the script will end without a way to put it in a loop.

**Comment by aes-gloot** [comment]
Oh, nice! Ok, I think the basic variant is working, but I do want to add the alias, if present...
```
<%
(async function() {
	const suggestInFolder = async (folder) => {
		const items = app.vault.getMarkdownFiles().filter(x => x.path.startsWith(folder));
		items.push({basename: ""})
		try {
		return (await tp.system.suggester((item) => item.basename, items)).basename;
		} catch(e) {
			return "";
		}
	}
 
	var p = await tp.system.prompt("Title");
	var s = `meeting:: ${p}\n`;
	while (p = await suggestInFolder("People/Professional")) {
		s += `- (contact:: [[${p}]])\n`
	}
	return s;
})()
%>
```

**Reply by Zachatoo** [reply]
Nice!

To get the alias, you'll need to pass a `TFile` to `app.metadataCache.getFileCache()` then parse out the alias in the frontmatter.

```js
<%
(async function() {
	const suggestInFolder = async (folder) => {
		const items = app.vault.getMarkdownFiles().filter(x => x.path.startsWith(folder));
		items.push({basename: ""})
		try {
                // Updated the below line to return the whole TFile instead of just the basename for each selection
		return await tp.system.suggester((item) => item.basename, items);
		} catch(e) {
			return "";
		}
	}
 
	var p = await tp.system.prompt("Title");
	var s = `meeting:: ${p}\n`;
	while (p = await suggestInFolder("People/Professional")) {
                 // Get aliases if available
                 const aliases = (await app.metadataCache.getFileCache(p)).frontmatter.aliases;
                 const alias = aliases?.length > 0 ? aliases[0] : null;
		s += `- (contact:: [[${p.basename}${alias ? `|${alias}` : ''}]])\n`
	}
	return s;
})()
%>
```

---

FYI, you don't have to use an IIFE to return and run JS in Templater, it supports [execution commands](https://silentvoid13.github.io/Templater/commands/execution-command.html) with top level await. This would be the more "Templater" way of doing it.

```js
<%*
const suggestInFolder = async (folder) => {
  const items = app.vault.getMarkdownFiles().filter(x => x.path.startsWith(folder));
  items.push({basename: ""})
  try {
  return await tp.system.suggester((item) => item.basename, items);
  } catch(e) {
    return "";
  }
}

var p = await tp.system.prompt("Title");
var s = `meeting:: ${p}\n`;
while (p = await suggestInFolder("People/Professional")) {
  const aliases = (await app.metadataCache.getFileCache(p)).frontmatter.aliases;
  const alias = aliases?.length > 0 ? aliases[0] : null;
  s += `- (contact:: [[${p.basename}${alias ? `|${alias}` : ''}]])\n`
}
tR += s;
-%>
```

**Comment by aes-gloot** [comment]
OK, what I've managed to build has issues, but I think the behavior of the selector is really what shuts this down... I can't get it search in the suggestions. I think this might be dead until I can actually just use that editor:insert-wikilink. (Also, I wasn't able to remove the IFFE ¯\_(ツ)_/¯ )

```javascript
(async function() {
var aliasLink = async (item) => {
  try {
    const aliases = (
      await app.metadataCache.getFileCache(item)
    ).frontmatter.aliases;
    return [item.basename]
    .concat(aliases)
    .map(a => [a, `${item.path}|${a}`])
  } catch {
    return [[item.basename, `${item.path}|${item.basename}`]]
  }
}

var suggestInFolder = async (folder) => {
  const items = app.vault.getMarkdownFiles().filter(x => x.path.startsWith(folder));
  const links = await Promise.all(items.map(aliasLink));
  console.log(links)
  try {
    return await tp.system.suggester(
      links.map(l => l[0]),
      links.map(l => l[1]),
    );
  } catch(e) {
    return "";
  }
}

var p = await tp.system.prompt("Title");
var s = `meeting:: ${p}\n`;
while (p = await suggestInFolder("People/Professional")) {
  s += `- (contact:: [[${p}]])\n`
}
return s;
})()
```

**Reply by Zachatoo** [reply]
I've modified my code to be able to search by aliases.

```js
<%*
const suggestInFolder = async (folder) => {
  const markdownFiles = app.vault
    .getMarkdownFiles()
    .filter((x) => x.path.startsWith(folder));

  const items = markdownFiles.reduce((acc, file) => {
    const aliases = app.metadataCache.getFileCache(file)?.frontmatter?.aliases;

    if (aliases && aliases.length > 0) {
      aliases.forEach((alias) => {
        acc.push({
          display: `${alias}\n${file.path}`,
          linkText: `${file.path}|${alias}`,
        });
      });
    } else {
      acc.push({
        display: `${file.basename}\n${file.path}`,
        linkText: `${file.path}|${file.basename}`,
      });
    }

    return acc;
  }, []);

  try {
    return await tp.system.suggester((item) => item.display, items);
  } catch (e) {
    return "";
  }
};

let p = await tp.system.prompt("Title");
let s = `meeting:: ${p}\n`;
while (p = await suggestInFolder("People/Professional")) {
  s += `- (contact:: [[${p.linkText}]])\n`;
}
tR += s;
-%>
```

That's so strange that you can't get rid of the IIFE, I only need an IFFE if I wrap in normal commands `<% ... %>` instead of execution commands `<%* ... %>`.
